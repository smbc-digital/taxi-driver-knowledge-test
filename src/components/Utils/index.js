import 'isomorphic-fetch'
import moment from 'moment'

export const formatAvailableAppointments = appointments => {
	return appointments.map(appointment => {
		return {
			date: appointment.date,
			displayDate: moment(appointment.date, 'DD/MM/YYYY').format('dddd D MMMM YYYY'),
			times: appointment.times.map(time => {
				const value = moment(`${appointment.date} ${time.startTime}`, 'DD/MM/YYYY HH:mm:ss').format('YYYY-MM-DD HH:mm:ss')
				return {
					startTime: time.startTime,
					id: value,
					value,
					label: moment(time.startTime, 'HH:mm:ss').format('H:mma'),
					name: 'testDate'
				}
			})
		}
	})
}

export const getAvailableAppointments = async (isResit, from, to) => {
	const request = {
		 from: from,
		 to: to,
		 isResit: isResit
	}

	try {
		const result = await fetch('/book-taxi-driver-knowledge-test/available-appointments', {
			method: 'POST',
			headers: {
				Accept: 'application/json; charset=utf-8',
				'Content-Type': 'application/json; charset=utf-8'
			},
			body: JSON.stringify(request)
		})

		const responseObject = await result.json()

		return {
			status: result.status,
			appointments: formatAvailableAppointments(responseObject)
		}
	} catch (error) {
		return {
			status: 500,
			appointments: []
		}
	}
}

export const reserveAppointment = async ({isResit, testDate, firstName, lastName, phoneNumber, emailAddress, address, testType, previousTestDate}) => {
	try {
		const request = {
			testDate, 
			isResit,
			name: `${firstName.value} ${lastName.value}`,
			address: address.value.selectedAddress || `${address.value.addressLine1} ${address.value.addressLine2} ${address.value.town} ${address.value.postcode}`,
			email: emailAddress.value,
			phoneNumber: phoneNumber.value,
			testType: testType.value,
			previousTestDate: previousTestDate.value
		}

		const result = await fetch('/book-taxi-driver-knowledge-test/pencil-an-appointment', {
			method: 'POST',
			headers: {
				Accept: 'application/json; charset=utf-8',
				'Content-Type': 'application/json; charset=utf-8'
			},
			body: JSON.stringify(request)
		})

		const responseObject = await result.json()

		return {
			status: result.status,
			bookingId: responseObject
		}
	} 
	catch (error) {
		return {
			status: 500
		}
	}
}

export const getPaymentUrl = async (bookingId, testDate) => {
	try {
		const result = await fetch('/book-taxi-driver-knowledge-test/generate-payment-url', {
				method: 'POST',
				headers: {
					Accept: 'application/json; charset=utf-8',
					'Content-Type': 'application/json; charset=utf-8'
				},
				body: JSON.stringify({
					bookingId, testDate
				})
			})

		const url = await result.text()

		return {
			status: result.status,
			url
		}
	} catch (error) {
		return {
			status: 500
		}
	}
}
