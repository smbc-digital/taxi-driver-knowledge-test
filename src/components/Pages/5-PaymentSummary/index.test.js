import React from 'react'
import Enzyme, { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
// import { getPageRoute } from '../../../helpers/pagehelper'
import * as submitUtils from '../../Utils'
import { PaymentSummary } from './index'
// import renderer from 'react-test-renderer'

Enzyme.configure({ adapter: new Adapter() })

describe('PaymentSummary', () => {
	beforeEach(() => {
		fetch.resetMocks()
	})

	it('should call push on submit with the correct next page url when response is 200', async () => {
		// Arrange
		const url = 'http://www.test.url/paymentTest'
		const data = {
			testDate: '26/02/2019',
			selectedAppointment:{
				date: 'test',
				times:{
					startTime: 'asd'
				}
			}
		}
		const history = { push: jest.fn() }
		window.location.assign = jest.fn()
		submitUtils.getPaymentUrl = jest.fn().mockReturnValue({ status: 200, url: url })

		const wrapper = mount(<PaymentSummary context={data} history={history} />)

		// Act
		await wrapper.find('form').simulate('submit')

		// Assert
		expect(window.location.assign).toHaveBeenCalledWith(url)
	})

	// it('should call push on submit with an error page when reponse is not 200', async () => {
	// 	// Arrange
	// 	const url = 'http://www.test.url/paymentTest'
	// 	const data = {
	// 		selectedAppointment:{
	// 			date: 'test',
	// 			times:{
	// 				startTime: 'asd'
	// 			}
	// 		}
	// 	}
	// 	const history = { push: jest.fn() }
	// 	window.location.assign = jest.fn()
	// 	submitUtils.getPaymentUrl = jest.fn().mockReturnValue({ status: 400, url: url })

	// 	const wrapper = mount(<PaymentSummary context={data} history={history} />)

	// 	// Act
	// 	await wrapper.find('form').simulate('submit')

	// 	// Assert
	// 	expect(history.push).toHaveBeenCalledWith(getPageRoute(7))
	// })

	// it('should not call push on submit when recaptcha is invalid', () => {
	// 	// Arrange
	// 	document.body.innerHTML = '<span id="displayRecaptcha" class="hidden">true</span>'
	// 	const data = {
	// 		displayRecaptcha: true,
	// 		selectedAppointment:{
	// 			date: 'test',
	// 			times:{
	// 				startTime: 'asd'
	// 			}
	// 		}
	// 	}

	// 	let history = { push: jest.fn() }
	// 	let wrapper = mount(<PaymentSummary context={data} history={history} />)

	// 	// Act
	// 	wrapper.find('button').simulate('click')

	// 	// Assert
	// 	expect(history.push).not.toBeCalled()
	// })

	// it('should set recaptchaValid in state to false', () => {
	// 	// Arrange
	// 	document.body.innerHTML = '<span id="displayRecaptcha" class="hidden">true</span>'
	// 	const data = {
	// 		displayRecaptcha: true,
	// 		selectedAppointment:{
	// 			date: 'test',
	// 			times:{
	// 				startTime: 'asd'
	// 			}
	// 		}
	// 	}

	// 	let history = { push: jest.fn() }
	// 	let wrapper = mount(<PaymentSummary context={data} history={history} />)

	// 	// Act
	// 	wrapper.instance().onChangeRecaptcha(undefined)

	// 	// Assert
	// 	expect(wrapper.state().recaptchaValid).toBe(false)
	// })

	// it('should set recaptchaValid in state to true', () => {
	// 	// Arrange
	// 	document.body.innerHTML = '<span id="displayRecaptcha" class="hidden">true</span>'
	// 	const data = {
	// 		displayRecaptcha: true,
	// 		selectedAppointment:{
	// 			date: 'test',
	// 			times:{
	// 				startTime: 'asd'
	// 			}
	// 		}
	// 	}

	// 	let history = { push: jest.fn() }
	// 	let wrapper = mount(<PaymentSummary context={data} history={history} />)

	// 	// Act
	// 	wrapper.instance().onChangeRecaptcha('testRecaptchaValue')

	// 	// Assert
	// 	expect(wrapper.state().recaptchaValid).toBe(true)
	// })

	// describe('snapshot', () => {
	// 	it('renders correctly', () => {
	// 		const context = {
	// 			whyMoreSpace: {
	// 				value: '',
	// 				isValid: false
	// 			},
	// 			whyMoreSpaceOther: {
	// 				value: '',
	// 				isValid: false
	// 			},
	// 			selectedAppointment:{
	// 				date: 'test',
	// 				times:{
	// 					startTime: 'asd'
	// 				}
	// 			}
	// 		}

	// 		const tree = renderer.create(<PaymentSummary context={context} />).toJSON()

	// 		expect(tree).toMatchSnapshot()
	// 	})
	// })
})
