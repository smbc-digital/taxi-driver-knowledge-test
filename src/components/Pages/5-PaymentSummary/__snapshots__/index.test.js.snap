// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentSummary snapshot renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <h1 />
  <h2>
    Your booking summary
  </h2>
  <p>
    Your booking to take the 
    Private Hire
     taxi driver knowledge test is at 
    8:45am
     on 
    Tuesday 2 April
    .
  </p>
  <p>
    You'll now be taken to our online system to make the payment.
  </p>
  <p>
    The cost is 
    <strong>
      Â£70
    </strong>
    .
  </p>
  <button
    className="button-primary"
    disabled={false}
    id="continue-to-payment"
    onClick={undefined}
    type="submit"
  >
    Continue to payment
  </button>
  <p>
    <a
      href="#"
      onClick={[Function]}
    >
      Previous
    </a>
  </p>
</form>
`;
